### PR TITLE
dual-kawase blur algorithm for new OpenGL backend

### PR DESCRIPTION
**Pull request by tryone144**
This PR implements the dual-filter kawase blur algorithm as [found in KWin](https://kwin.kde.narkive.com/aSqRYYw7/d9848-updated-the-blur-method-to-use-the-more-efficient-dual-kawase-blur-algorithm).

demo

Usable with `--blur-method dual_kawase --blur-strength 7`.

**~~This is currently work-in-progress and open for comments. DO NOT MERGE.~~**
Should be ready to merge now.

Previous discussion: https://github.com/yshui/picom/issues/32
See also: https://github.com/yshui/picom/pull/361

EDIT: Sorry, clicked submit too fast...

### CODE REVIEW

File: src/backend/backend_common.c
Comment (tryone144) [inline src/backend/backend_common.c:441]: This could be possible with a lookup-table and a for-loop as well. Not sure what might be cleaner.

File: src/backend/backend_common.c
Comment (tryone144) [inline src/backend/backend_common.c:454]: Whatever I tried doing there, this is too complex. At least the `exp2f` part can be expressed by bit-shifts.
This should calculate the `expand` size to cover the sample offset in the smallest blur-texture.

File: src/config.c
Comment (tryone144) [inline src/config.c:92]: As my first implementation for compton used the slightly incorrect name "kawase" for this parameter it might be a good idea to accept it as well and output a deprecation message.

File: src/config_libconfig.c
Comment (tryone144) [inline src/config_libconfig.c:631]: This is a duplicate check. Checked again in `src/options.c`, which should be the correct place.

File: src/config_libconfig.c
Comment (tryone144) [inline src/config_libconfig.c:647]: This is a duplicate check. Checked again in `src/options.c`, which should be the correct place.

**Comment by tryone144** [comment]
Apart from the comments above:
- Commit https://github.com/yshui/picom/pull/382/commits/46b48ebb4ad0183f58e779e0753c7c9fb9ebae91 has to be removed prior to merging
- Commit https://github.com/yshui/picom/pull/382/commits/209d9b6558e430033d7ccd91e8657aea1670d1c0 might be better off in a separate PR. Merging this beforehand reduced the complexity of the blur-code somewhat.

File: src/backend/backend_common.c
Comment (yshui) [reply src/backend/backend_common.c:441]: I think a lookup-table would be better.

File: src/config.c
Comment (yshui) [reply src/config.c:92]: Agreed

File: src/backend/backend_common.c
Comment (mighty9245) [reply src/backend/backend_common.c:454]: `(256 - (int)(256.0f - params->offset))` - can't this be replaced with `(int)ceil(params->offset)`?

Comment (tryone144) [reply src/backend/backend_common.c:454]: This looks indeed like a manual `ceil()` operation. The formula behind should be something around `2 * ceil(sample offset) * max scaling factor + 1` to cover the required sample area.

Comment (mighty9245) [reply src/backend/backend_common.c:454]: i updated my comment with my variant of this formula that seems to produce the same result but looks simpler

**Comment by Shadorain** [comment]
Hey guys! Yall's fork is top, I just got into tiling window managers and wouldnt survive without your fork. I was wondering if it is possible (maybe I missed something even after hours of searching) to have blur set up to be across all windows, but then increase the blur strength on the focused window, just an idea that I feel would cap my theme! Thanks!

**Comment by tryone144** [comment]
@Shadorain having different strength-levels similar to active/inactive opacity should be possible with a second set of parameters. But then you could also request per-window rules of blur-strength... I think it should be priority to get blur-algo in general merged first. :wink:

Keep in mind, that strength-fading with this algorithm requires some hackery and will never be smooth due to the discrete nature of the internal sample-iterations. See https://github.com/yshui/picom/issues/32#issuecomment-439293203 for an explanation and https://github.com/tryone144/compton/issues/2 for some hacked samples.

**Comment by Shadorain** [comment]
> @Shadorain having different strength-levels similar to active/inactive opacity should be possible with a second set of parameters. But then you could also request per-window rules of blur-strength... I think it should be priority to get blur-algo in general merged first. wink
> 
> Keep in mind, that strength-fading with this algorithm requires some hackery and will never be smooth due to the discrete nature of the internal sample-iterations. See [#32 (comment)](https://github.com/yshui/picom/issues/32#issuecomment-439293203) for an explanation and [tryone144/compton#2](https://github.com/tryone144/compton/issues/2) for some hacked samples.

to be honest, I'm not sure what I did to make it this way, but when i remade my entire configuration file by hand just now, somehow (not even trying) I was able to get a very strange effect that I'm not too sure if its a bug or not
But it kind of does what I wanted in my previous post (which doesn't really make the most sense lol) I'll upload my conf file, it's very clean with alot less lines which makes it easier to read and edit

Thank you for the quick reply! I will see what i can do about your comment too, how i can implement it

File: src/backend/gl/gl_common.c
Comment (tryone144) [inline src/backend/gl/gl_common.c:1336]: Since the VAOs for the window quads supply the texture coordinates in actual pixel dimensions, we have to transform them back to normalized texture coordinates. Pixel coordinates are actually useful for the simplistic kernel blur with `texelFetch()`. The kawase kernels however relies on the linear filtering of `texture2D()` which works with normalized coordinates.
Could we benefit from supplying normalized texture coords in `x_rect_to_coords()` with a flag in case we want to blur with dual-kawase? We should be able to skip updating the `texture_size` uniform with every iteration. (And `scale` would not be needed anymore in the fragment shaders.)

**Comment by mighty9245** [comment]
hey, @yshui, since the pull request is here (though it still requires some work, but i don't think there is much to do), may i ask you: are you planning to merge it before releasing v8? would be amazing to have dual kawase blur in v8 (and rounded corners in v9, hehe) :)

**Comment by tryone144** [comment]
I wouldn't want to rush things, especially in light of https://github.com/yshui/picom/issues/385. Should be fine with releasing v8 first and then add these visual features in v9.

Except for the small changes outlined above, this code should be feature complete and seems to run without issues. The thing I am currently kinda stuck on is how to handle https://github.com/yshui/picom/pull/349 as these changes have some implications on the rendering logic.

**Comment by Philipp-M** [comment]
I've noted a small visual issue, if windows are  next to each other.
Windows are borderless in the following example and have the same color and transparency.
I've used the default blur strength 5, no shadows, with only a wallpaper behind the windows.
screenshot-2020-04-22_17:41:21

**Comment by tryone144** [comment]
> I've noted a small visual issue, if windows are next to each other.

This is to be expected and due to the way windows are rendered. Each window is rendered one by one and the current result is used for blur calculation (i.e. the top-most window includes the already rendered windows below into its calculations as one would expect).
The two windows in your example aren't on-top of each other strictly speaking, but are still rendered in order (most likely the active window as last).

It would be possible to only take the pixel directly below the window into account for bluring, but this will introduce artifacts when pixels directly at the window border change — mostly noticable when moving a floating window. See [this comment](https://github.com/tryone144/compton/commit/a7c5c43d3229fab3bd18dabae691836edd59ad9b#commitcomment-28413876) for a more in-depth explanation.

This should however be fixable by something like https://github.com/yshui/picom/issues/265, although the implementation in https://github.com/yshui/picom/commit/6a3d1354bed9d4c33227944a2331b0e5713668d3 seems to clip to little leading to artifacts on the edge of the windows as there are still parts of the underlying window sampled. This is a different issue though and has to be fixed separately.

**Comment by yshui** [comment]
@tryone144 The clipping added in #265 clips the window directly under the transparent window's body, so it's expected to have windows around the transparent window included in the blur. I agree it's not pretty. We could probably use `root_pixmap` instead of the target buffer for blur when `transparent-clipping` is enabled, should be pretty easy to do.

**Comment by tryone144** [comment]
> We could probably use `root_pixmap` instead of the target buffer for blur when `transparent-clipping` is enabled, should be pretty easy to do.

This sounds like a good solution. Actually, that's exactly what I thought you would implement when I read the posts. :laughing:
But I agree, fix the problems step-by-step.

**Comment by Philipp-M** [comment]
Another idea I just had regarding this issue should still allow visual overlapping of windows.

I guess currently the sorting of the windows is done by the window manager, and in this order the windows are drawn by the compositor.
I'm not too much in the internals of window managing/compositing with X, so I'm not sure if it's working:

Resort the window draw order, allowing multiple windows drawn at once / with the same background framebuffer based on how they're overlapping (maintaining the original order).
This should be possible with a simple rectangle intersection (or maybe even a fancier but more costly pixel-based intersection for non-rectangular windows probably using the gpu if possible)
If windows are not overlapping (like in the example above) or have the same windows beneath them, they are drawn in the same pass with the same background framebuffer.

This could possibly also solve problems with shadows overlapping neighboring windows (I'm pretty sure it's the same issue)

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c:1336]: `texelFetch()` does not implement any repeat/clamp logic like `texture2D()` does. This would have to be manually implemented in the shader requiring the actual source texture dimensions. Currently this leads to darkened blur around screen borders in the kernel blur implementation.

I'd suggest adopting the kernel blur shader to use `texture2D()` as well with correctly setup textures and provide normalized coordinates in `x_rect_to_coords()` when requested.

**Comment by mighty9245** [comment]
image
looks like we need an ability to turn off blurring for a wintype?

**Comment by tryone144** [comment]
&lt;rant&gt;
I really hate all those tooltips menus and splashscreens which think rendering their own shadow is an incredibly great idea... And typically these are the ones that don't set appropriate wintypes. :facepalm:
&lt;/rant&gt;

Extending the `blur-exclude` to wintype options seems like a good idea, though as a separate change.

**Comment by mighty9245** [comment]
yeah, i had this problem with all tooltip menus initially, but i was too lazy to dig in, so i just set `blur-background-exclude` to match all windows except the terminal windows, but terminal also has tooltip menus...

idk what you meant by "extending the `blur-exclude` to wintype options", but we already have `wintypes` thing that allows to configure different things like shadows, focus for different wintypes and i want to have `blur` or `blur-background` option there, like:
```
wintypes: {
  tooltip = {
    blur-background = false;
  };
};
```

if we are going for this, would be good to implement this in a separate pull request as well and merge before this.

**Comment by tryone144** [comment]
> not sure what you meant with "extending the `blur-exclude` to wintype options", but we already have `wintypes` thing that allows to configure different things like shadows, focus for different wintypes and i want to have `blur` or `blur-background` option there
> [snip]

Exactly this. :laughing:

**Comment by tryone144** [comment]
Applied the changes discussed above and squashed some unnecessary commits.

> * Commit [46b48eb](https://github.com/yshui/picom/commit/46b48ebb4ad0183f58e779e0753c7c9fb9ebae91) has to be removed prior to merging

Done.

> * Commit [209d9b6](https://github.com/yshui/picom/commit/209d9b6558e430033d7ccd91e8657aea1670d1c0) might be better off in a separate PR. Merging this beforehand reduced the complexity of the blur-code somewhat.

Merged with #388 and #389.

> Currently this leads to darkened blur around screen borders in the kernel blur implementation.
> I'd suggest adopting the kernel blur shader to use texture2D() as well with correctly setup textures and provide normalized coordinates in x_rect_to_coords() when requested.

Will take a look into this next and then rebase on top of that PR.

**Comment by vn-ki** [comment]
@tryone144 

```
[ 05/01/2020 14:52:44.508 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0040032e (vn-ki - File Manager)
```

Also segfaults when closing a window sometimes.

<details>
<summary>My config</summary>

```
# ROunded corners
corner-radius = 10;
# Shadow
shadow = true;
#shadow-radius = 30;
#shadow-offset-x = -30;
#shadow-offset-y = -30;
shadow-radius = 20;
shadow-offset-x = -20;
shadow-offset-y = -20;
log-level = "warn";
# log-file = "/path/to/your/log/file";
shadow-opacity = 0.8;
# shadow-red = 0.0;
# shadow-green = 0.0;
# shadow-blue = 0.0;
shadow-exclude = [
	"_COMPTON_NO_SHADOW@:c = 1",
    # for zoom screen share
    "name = 'cpt_frame_window'",

    #"class_g = 'Google-chrome-beta'",
#	"name = 'Notification'",
#	"class_g = 'Conky'",
#	"class_g ?= 'Notify-osd'",
#	"class_g = 'Cairo-clock'",
#	"_GTK_FRAME_EXTENTS@:c"
];
# shadow-exclude = "n:e:Notification";
# shadow-exclude-reg = "x10+0+0";
# xinerama-shadow-crop = true;

# Opacity
#inactive-opacity = 0.9;
# active-opacity = 0.8;
frame-opacity = 0.7;
inactive-opacity-override = false;
 inactive-dim = 0.2;
# inactive-dim-fixed = true;
#blur-background = true;
#blur-background-frame = true;
blur-method = "dual_kawase";
blur-strength=10;
# blur-kern = "3x3box";
# blur-kern = "5,5,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1";
blur-background-fixed = true;
blur-background-exclude = [
    # "window_type = 'dock'",
    "window_type = 'desktop'",
    "_GTK_FRAME_EXTENTS@:c"
];
opacity-rule = [ 
    "80:class_g = 'URxvt'",
    # "70:class_i = 'music.youtube.com'",
    # "90:class_i = 'discordapp.com__activity'",
];

# Fading
fading = true;
fade-delta = 5;
fade-in-step = 0.03;
fade-out-step = 0.03;
# no-fading-openclose = true;
# no-fading-destroyed-argb = true;
fade-exclude = [
#"class_g = 'Google-chrome-beta'",
];

# Other
#backend = "xr_glx_hybrid";
backend = "glx";

#backend = "xr_glx_hybrid";
#vsync-use-glfinish=true;

mark-wmwin-focused = true;
mark-ovredir-focused = true;
# use-ewmh-active-win = true;
detect-rounded-corners = true;
detect-client-opacity = true;
#refresh-rate = 60;
vsync = true;
#vsync = "opengl-swc";
#sw-opti = true;
#unredir-if-possible = true;
#unredir-if-possible-delay = 5000;
# unredir-if-possible-exclude = [ ];
focus-exclude = [ 
"class_g = 'Cairo-clock'", 
"class_g = 'Google-chrome-beta'" 
];
detect-transient = true;
detect-client-leader = true;
invert-color-include = [ ];
# resize-damage = 1;

# GLX backend
glx-no-stencil = true;
#glx-no-rebind-pixmap = true;
#xrender-sync-fence = true;
#use-damage = true;

# Window type settings
wintypes:
{
  tooltip = { fade = true; shadow = true; opacity = 0.75; focus = true; full-shadow = false; };
  dock = { shadow = false; }
  dnd = { shadow = false; }
  menu = { shadow = false; }
  #popup_menu = { opacity = 0.8; }
  popup_menu = {shadow = false; }
  #notify = {shadow = false; }
  dropdown_menu = { shadow = false; }
  # toolbar = { shadow = false; }
  # utility = { shadow = false; }
  # toolbar = { shadow = false; }
  # splash = { shadow = false; }
  dialog = { shadow = false; }
};

```

</details>

**Comment by tryone144** [comment]
> ```
> [ 05/01/2020 14:52:44.508 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0040032e (vn-ki - File Manager)
> ```
> 
> Also segfaults when closing a window sometimes.

This issue is not directly related to this PR and happens on `next` as well. @vn-ki please open a separate issue for this. At least a stack trace would be pretty useful to further investigate this.
I have run into this several times now while testing (both `glx` and `xrender` backend) and it seems to be related to some state housekeeping when windows rapidly map/unmap (really bad case: Memory pressure and "fading" vlc controls in fullscreen).

**Comment by yshui** [comment]
Perhaps the same crash as #390?

**Comment by tryone144** [comment]
> Perhaps the same crash as #390?

Yep, looks like the ones I encountered (triggered in `paint_all_new()`). That isn't the only trigger I've found though.

**Comment by yshui** [comment]
@tryone144 Can you open an issue with stacktrace? We can close it later if it's duplicated.

**Comment by arking7** [comment]
Only some of the apps have the blur effect on the title bar, the rest is still transparent, how can I make all the title bars blur?
Screenshot_2020-05-15_15-59-46

**Comment by mighty9245** [comment]
@arking7, set `blur-background-frame` to `true`.

**Comment by arking7** [comment]
@mighty9245 Thank you very much, it works now.

**Comment by arking7** [comment]
Now there's a new problem, the blur effect bleeds out of the frame after I enable "blur-background-frame".😢
image

**Comment by tryone144** [comment]
> Now there's a new problem, the blur effect bleeds out of the frame after I enable "blur-background-frame".

Does this only happen with the `dual_kawase` blur method? If so, please open a new issue with more details over at my repo. If it also happens with other blur methods / backends / etc., it might be better to open a new issue with more details about your system and how to reproduce the problem.

**Comment by yshui** [comment]
@arking7 That looks like a problem with the general blur code, and it might have been fixed. If you are already using v8, please open an issue; otherwise, can you upgrade and try again?

**Comment by arking7** [comment]
> @arking7 That looks like a problem with the general blur code, and it might have been fixed. If you are already using v8, please open an issue; otherwise, can you upgrade and try again?

I have tried with stable/8 if I understand v8 right, and same problem still exists with "gaussian" method.
blur

**Comment by yshui** [comment]
@arking7 OK, I will look into it. Please open an issue if you can.

**Comment by arking7** [comment]
> @arking7 OK, I will look into it. Please open an issue if you can.

Thanks for you reply.

I have opened a new issue #413, if you need more details just let me know.

**Comment by tryone144** [comment]
Running flawlessly on my setup since the last rebase (apart from the unrelated assertion failures in the experimental backends) and I haven't gotten any complaints, so it seems to be good for merging.

If there is enough demand, I can try to cleanly backport the dual_kawase code to the legacy backend as well.

**Comment by yshui** [comment]
I will try to review this ASAP.

P.S. @tryone144, may I interest you in becoming a collaborator of the picom project? (In other words, can you help me review PRs, handle issue and maybe fix bugs for picom, _please_?)

**Comment by tryone144** [comment]
> P.S. @tryone144, may I interest you in becoming a collaborator of the picom project? (In other words, can you help me review PRs, handle issue and maybe fix bugs for picom, _please_?)

I'll gladly help out with handling issues and PRs. :smile:
I only have limited time to spend though, so I can't promise too much from my side. If this is okay with you, feel free to hit me up with more details.

**Comment by yshui** [comment]
@tryone144 It's still better than just having me.

Basically I am going to add you as a collaborator on this project, so you will be able to push to this repository, create new branches here, and approve pull requests.

Although you will be able to push directly to the `next` branch in theory. For more complex changes, I still want you to open PRs so more people can look at it (I will do the same). For trivial changes, you can just push to `next`. I'll leave it to you (and to me) to decide what is "trivial", and hopefully we will hash out the details as time progresses.

**Comment by tryone144** [comment]
Sounds good to me. :)

File: src/backend/gl/gl_common.c
Comment (yshui) [inline src/backend/gl/gl_common.c:49]: Can you add a comment about what these variables are for? I couldn't figure it out just by the name.

**Review by yshui — CHANGES_REQUESTED** [review]
Still going through the last commit.

A small nit.

File: src/backend/gl/gl_common.c
Comment (yshui) [inline src/backend/gl/gl_common.c:639]: Same thing as `width >> iterations`, `height >> iterations`, I believe?

File: src/backend/gl/gl_common.c
Comment (yshui) [inline src/backend/gl/gl_common.c:643]: What does this check for?

File: src/backend/gl/gl_common.c
Comment (yshui) [inline src/backend/gl/gl_common.c:764]: Same as above, `(bctx->fb_width - 1) >> (i + 1)`, etc.

File: src/backend/gl/gl_common.c
Comment (yshui) [reply src/backend/gl/gl_common.c:639]: Curious: does this mean smaller windows will get weaker blur?

**Comment by yshui** [comment]
LGTM, only found some nits.

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c:49]: Will do. I'll also have to take a look again at those new `clang-tidy` warnings.

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c:639]: > Same thing as `width >> iterations`, `height >> iterations`, I believe?

Now that you say it. Should yield the same results, yes. ^^
And yes, we skip those iterations where we would render less than a full pixel in either width or height.

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c:643]: Trying to explain `clang-tidy` the scale factor won't be too large [over here](https://github.com/yshui/picom/blob/b6987352499bd23b8eae73553e64aa9169bcb6e6/src/backend/gl/gl_common.c#L734). But I am not quite comfortable with how it turned out.

File: src/backend/gl/gl_common.c
Comment (yshui) [reply src/backend/gl/gl_common.c:49]: Note with the new `.clang-tidy`, it doesn't run clang-analyzer checks anymore. But meson lets you run them easily with `ninja scan-build`

File: src/backend/gl/gl_common.c
Comment (yshui) [reply src/backend/gl/gl_common.c:639]: Hmm, I wonder if it's going to look weird.

Not relevant to this PR, but I am thinking if we could always blur the whole screen, cache the result, and clip out the needed parts when rendering blurred background. This way we could get consistent blur strength, but I am not sure about performance.

File: src/backend/gl/gl_common.c
Comment (yshui) [reply src/backend/gl/gl_common.c:643]: You can have a `scale` variable, and double it every iteration. clang-tidy probably won't complain about that.

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c:639]: Windows have to be really tiny to notice. This was implemented mainly for performance reasons but you will get a noticeable change in the blur itself when you cross the threshold. We should be fine with leaving this one out (not sure about the OpenGL spec but my nvidia hardware doesn't give me any glitches when drawing less than a single pixel wide lines, probably because we already sample more due to `resize_width`/`resize_height` being added).

Regarding full-screen blur: Unless we are using `--transparent-clipping`, we would need to blur the whole screen multiple times (at least for each "stacking layer"). Not sure if that logic would be more performant than rendering each window separately. Restricting blur to xinerama-regions/xrandr-outputs should be easier though. Definitely stuff for later.

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c:52]: Please take a look at the updated comments. We could probably get away with leaving `fb_*` out but this would require some assumptions about the `texture_sizes` array (at least one entry, first one should be equal to the current `fb_*` value) that aren't obvious and not given at the moment.

`texture_sizes` is only really needed by the dual-filter kawase algorithm, as for the kernel blur these are equal to `fb_width`/`fb_height`. These can probably also be calculated on the fly, but I think it's cleaner to explicitly cache them than to calculate them at multiple places.

File: src/backend/gl/gl_common.c
Comment (tryone144) [reply src/backend/gl/gl_common.c]: Adding another variable to keep track of the current scaling factor worked out nicely.

File: src/backend/gl/gl_common.c
Comment (yshui) [reply src/backend/gl/gl_common.c]: > We should be fine with leaving this one out 

Keep it in, even without the check, the blur is going to be weaker anyway, right?

Comment (tryone144) [reply src/backend/gl/gl_common.c]: > Keep it in, even without the check, the blur is going to be weaker anyway, right?

No, it should be consistent regardless of the window size.

I've done some testing and at the highest level (i.e. 5 iterations), the window has to be smaller than 32px in either direction for that to be noticable. With the check removed I could not produce any visual glitches with windows only a few pixels wide/tall. But I am pretty sure that is because we compensate with `resize_width` and `resize_height` anyway to have enough pixels outside the window for the last iteration.

I've implemented this check back in the old compton codebase without the expand/resize logic and where textures were allocated for each blur-call where less than 1px textures were a problem.

Comment (yshui) [reply src/backend/gl/gl_common.c]: I see, that makes sense then.

**Comment by yshui** [comment]
We can finally merge this after so long. Feels like there needs to be some celebration of some sort.

**Comment by yshui** [comment]
Also, huge thanks to @tryone144 for their amazing work, patience and perseverance.

**Pull request merged**
